tracer/src/Datadog.Trace/Agent/DiscoveryService/DiscoveryService.cs

```diff
@@ -17,144 +20,232 @@ namespace Datadog.Trace.Agent.DiscoveryService
+                catch (OperationCanceledException)
                 {
-                    return false;
                 }
-
-                sleepDuration = Math.Min(sleepDuration * 2, sleepMaxDuration);
             }
 
-            return false;
-        }
+            Log.Debug("Discovery service exiting");
 
-        /// <summary>
-        /// Sets the DiscoveryService instance.
-        /// Intended use is for unit testing
-        /// </summary>
-        /// <param name="discoveryService">DiscoveryService instance</param>
-        internal static void SetDiscoveryService(DiscoveryService discoveryService)
-        {
-            lock (GlobalLock)
-            {
-                Instance = discoveryService;
-            }
+            int GetNextSleepDuration(int? previousDuration) =>
+                previousDuration is null ? _initialRetryDelayMs : Math.Min(previousDuration.Value * 2, _maxRetryDelayMs);
         }
 
         private void ProcessDiscoveryResponse(string content)
         {
             var jObject = JsonConvert.DeserializeObject<JObject>(content);
-            AgentVersion = jObject["version"]?.Value<string>();
-
-            var discoveredEndpoints = (jObject["endpoints"] as JArray)?.Values<string>().ToArray();
-            if (discoveredEndpoints == null || discoveredEndpoints.Length == 0)
+            if (jObject is null)
             {
-                return;
+                throw new Exception("Error deserializing discovery response: response was null");
             }
 
-            ConfigurationEndpoint = SupportedConfigurationEndpoints
-               .FirstOrDefault(
-                    supportedEndpoint => discoveredEndpoints.Any(
-                        endpoint => endpoint.Trim('/').Equals(supportedEndpoint, StringComparison.OrdinalIgnoreCase)));
+            var agentVersion = jObject["version"]?.Value<string>();
+            var clientDropP0 = jObject["client_drop_p0s"]?.Value<bool>() ?? false;
+
+            var discoveredEndpoints = (jObject["endpoints"] as JArray)?.Values<string>().ToArray();
+            string? configurationEndpoint = null;
+            string? debuggerEndpoint = null;
+            string? statsEndpoint = null;
+            string? dataStreamsMonitoringEndpoint = null;
 
-            DebuggerEndpoint = SupportedDebuggerEndpoints
-               .FirstOrDefault(
-                    supportedEndpoint => discoveredEndpoints.Any(
-                        endpoint => endpoint.Trim('/').Equals(supportedEndpoint, StringComparison.OrdinalIgnoreCase)));
+            if (discoveredEndpoints is { Length: > 0 })
+            {
+                foreach (var discoveredEndpoint in discoveredEndpoints)
+                {
+                    var endpoint = discoveredEndpoint?.Trim('/');
+                    if (endpoint is null)
+                    {
+                        continue;
+                    }
+
+                    // effectively a switch, but case insensitive
```

Comment:
I think we can safely assume the endpoint we receive from agent's `info` is always lowercase.

Reply:
I'm happy to change it, I only used case insensitive because that's what was used before 😄

Reply:
so... I didn't expect it... but the `switch` is, IMO, uglier 😂 I might leave it as it-is with the excuse of case sensitivity 😉 

```csharp
switch (endpoint)
{
    case SupportedDebuggerEndpoint:
        debuggerEndpoint = endpoint;
        break;
    case SupportedConfigurationEndpoint:
        configurationEndpoint = endpoint;
        break;
    case SupportedStatsEndpoint:
        statsEndpoint = endpoint;
        break;
    case SupportedDataStreamsEndpoint:
        dataStreamsMonitoringEndpoint = endpoint;
        break;
    default:
        Log.Debug("Unused discovered endpoint {endpoint}", endpoint);
        break;
}
```